File: server/services/pds-data-fetcher.ts
/**
 * PDS Data Fetcher Service
 * 
 * Fetches missing data from source PDS when Redis entries are incomplete
 * due to missing referenced users or posts.
 */

import { didResolver } from "./did-resolver";
import { pdsClient } from "./pds-client";
import { storage } from "../storage";
import { eventProcessor } from "./event-processor";
import { CID } from 'multiformats/cid';
import * as Digest from 'multiformats/hashes/digest';

interface PDSDataFetchResult {
  success: boolean;
  data?: any;
  error?: string;
}

interface IncompleteEntry {
  type:
    | 'user'
    | 'post'
    | 'like'
    | 'repost'
    | 'follow'
    | 'list'
    | 'listitem'
    | 'feedgen'
    | 'starterpack'
    | 'labeler'
    | 'record'; // generic record fetch by URI
  did: string;
  uri?: string;
  missingData?: any;
  retryCount: number;
  lastAttempt: number;
}

export class PDSDataFetcher {
  private readonly MAX_RETRY_ATTEMPTS = 3;
  private readonly RETRY_DELAY_MS = 30000; // 30 seconds
  private readonly FETCH_TIMEOUT_MS = 10000; // 10 seconds
  private incompleteEntries = new Map<string, IncompleteEntry>();
  private isProcessing = false;

  constructor() {
    // Start periodic processing of incomplete entries
    this.startPeriodicProcessing();
  }

  /**
   * Mark an entry as incomplete and needing data fetch
   */
  markIncomplete(
    type:
      | 'user'
      | 'post'
      | 'like'
      | 'repost'
      | 'follow'
      | 'list'
      | 'listitem'
      | 'feedgen'
      | 'starterpack'
      | 'labeler'
      | 'record',
    did: string,
    uri?: string,
    missingData?: any,
  ) {
    const key = `${type}:${did}:${uri || ''}`;
    const existing = this.incompleteEntries.get(key);
    
    if (existing) {
      existing.retryCount++;
      existing.lastAttempt = Date.now();
      existing.missingData = missingData || existing.missingData;
    } else {
      this.incompleteEntries.set(key, {
        type,
        did,
        uri,
        missingData,
        retryCount: 0,
        lastAttempt: Date.now()
      });
    }

    console.log(`[PDS_FETCHER] Marked ${type} as incomplete: ${did} (${this.incompleteEntries.size} total incomplete)`);
  }

  /**
   * Start periodic processing of incomplete entries
   */
  private startPeriodicProcessing() {
    // Process every 30 seconds
    setInterval(async () => {
      if (this.isProcessing) return;
      
      try {
        await this.processIncompleteEntries();
      } catch (error) {
        console.error('[PDS_FETCHER] Error processing incomplete entries:', error);
      }
    }, 30000);
  }

  /**
   * Process all incomplete entries
   */
  private async processIncompleteEntries() {
    if (this.incompleteEntries.size === 0) return;
    
    this.isProcessing = true;
    console.log(`[PDS_FETCHER] Processing ${this.incompleteEntries.size} incomplete entries...`);

    const entries = Array.from(this.incompleteEntries.entries());
    let processed = 0;
    let success = 0;

    for (const [key, entry] of entries) {
      try {
        // Skip if too many retries
        if (entry.retryCount >= this.MAX_RETRY_ATTEMPTS) {
          console.warn(`[PDS_FETCHER] Max retries exceeded for ${key}, removing`);
          this.incompleteEntries.delete(key);
          continue;
        }

        // Skip if recently attempted
        const timeSinceLastAttempt = Date.now() - entry.lastAttempt;
        if (timeSinceLastAttempt < this.RETRY_DELAY_MS) {
          continue;
        }

        const result = await this.fetchMissingData(entry);
        if (result.success) {
          success++;
          this.incompleteEntries.delete(key);
          console.log(`[PDS_FETCHER] Successfully fetched data for ${key}`);
        } else {
          console.warn(`[PDS_FETCHER] Failed to fetch data for ${key}: ${result.error}`);
        }
        
        processed++;
      } catch (error) {
        console.error(`[PDS_FETCHER] Error processing ${key}:`, error);
      }
    }

    console.log(`[PDS_FETCHER] Processed ${processed} entries, ${success} successful, ${this.incompleteEntries.size} remaining`);
    this.isProcessing = false;
  }

  /**
   * Fetch missing data from PDS
   */
  private async fetchMissingData(entry: IncompleteEntry): Promise<PDSDataFetchResult> {
    try {
      // Resolve DID to PDS endpoint
      const pdsEndpoint = await didResolver.resolveDIDToPDS(entry.did);
      if (!pdsEndpoint) {
        return {
          success: false,
          error: `Could not resolve PDS endpoint for DID: ${entry.did}`
        };
      }

      console.log(`[PDS_FETCHER] Fetching data for ${entry.type} ${entry.did} from ${pdsEndpoint}`);

      switch (entry.type) {
        case 'user':
          return await this.fetchUserData(entry.did, pdsEndpoint);
        case 'post':
          return await this.fetchPostData(entry.did, entry.uri!, pdsEndpoint);
        case 'list':
        case 'listitem':
        case 'feedgen':
        case 'starterpack':
        case 'labeler':
        case 'record':
          return await this.fetchRecordByUri(entry.uri!, pdsEndpoint);
        case 'like':
        case 'repost':
        case 'follow':
          // Ensure the actor exists first; referenced subject fetch is handled via 'post'/'record' marks
          return await this.fetchUserData(entry.did, pdsEndpoint);
        default:
          return {
            success: false,
            error: `Unknown entry type: ${entry.type}`
          };
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Fetch user data from PDS
   */
  private async fetchUserData(did: string, pdsEndpoint: string): Promise<PDSDataFetchResult> {
    try {
      // Fetch the profile record directly from PDS using com.atproto.repo.getRecord
      // This is a PDS endpoint and doesn't require authentication
      const profileResponse = await fetch(
        `${pdsEndpoint}/xrpc/com.atproto.repo.getRecord?repo=${did}&collection=app.bsky.actor.profile&rkey=self`,
        {
          headers: { 'Accept': 'application/json' },
          signal: AbortSignal.timeout(this.FETCH_TIMEOUT_MS)
        }
      );

      if (!profileResponse.ok) {
        return {
          success: false,
          error: `Profile fetch failed: ${profileResponse.status}`
        };
      }

      const response = await profileResponse.json();
      const profile = response.value; // The actual profile record is in .value
      
      if (profile) {
        // Resolve handle from DID
        const handle = await didResolver.resolveDIDToHandle(did);
        
        // Extract avatar and banner CIDs from blob references (same logic as event processor)
        const extractBlobCid = (blob: any): string | null => {
          if (!blob) return null;
          
          if (typeof blob === 'string') {
            return blob === 'undefined' ? null : blob;
          }
          
          if (blob.ref) {
            if (typeof blob.ref === 'string') {
              return blob.ref !== 'undefined' ? blob.ref : null;
            }
            
            if (blob.ref.$link) {
              return blob.ref.$link !== 'undefined' ? blob.ref.$link : null;
            }
            
            // Binary CID object from PDS
            if (blob.ref.code !== undefined && blob.ref.multihash) {
              try {
                if (typeof blob.ref.toString === 'function' && blob.ref.toString !== Object.prototype.toString) {
                  const cidString = blob.ref.toString();
                  return cidString !== 'undefined' ? cidString : null;
                }
                
                const mh = blob.ref.multihash;
                const digest = mh.digest;
                
                let digestBytes: Uint8Array;
                if (digest && typeof digest === 'object' && !ArrayBuffer.isView(digest)) {
                  const size = mh.size || Object.keys(digest).length;
                  digestBytes = new Uint8Array(size);
                  for (let i = 0; i < size; i++) {
                    digestBytes[i] = digest[i];
                  }
                } else if (ArrayBuffer.isView(digest)) {
                  digestBytes = new Uint8Array(digest.buffer, digest.byteOffset, digest.byteLength);
                } else {
                  return null;
                }
                
                const multihashDigest = Digest.create(mh.code, digestBytes);
                const cidObj = CID.create(blob.ref.version || 1, blob.ref.code, multihashDigest);
                return cidObj.toString();
              } catch (error) {
                console.error('[PDS_FETCHER] Error converting binary CID:', error);
                return null;
              }
            }
          }
          
          if (blob.cid) {
            return blob.cid !== 'undefined' ? blob.cid : null;
          }
          
          return null;
        };
        
        const avatarCid = extractBlobCid(profile.avatar);
        const bannerCid = extractBlobCid(profile.banner);
        
        // Update user with full profile data
        await storage.updateUser(did, {
          handle: handle || did,
          displayName: profile.displayName || null,
          description: profile.description || null,
          avatarUrl: avatarCid,
          bannerUrl: bannerCid,
        });
        
        console.log(`[PDS_FETCHER] Updated user ${did} - handle: ${handle || did}, avatar: ${avatarCid ? 'YES' : 'NO'}, banner: ${bannerCid ? 'YES' : 'NO'}`);
        
        // Flush any pending operations for this user
        await eventProcessor.flushPendingUserOps(did);
        await eventProcessor.flushPendingUserCreationOps(did);
        
        return {
          success: true,
          data: { did, handle: handle || did, profile }
        };
      } else {
        return {
          success: false,
          error: 'No profile record found'
        };
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Fetch post data from PDS
   */
  private async fetchPostData(authorDid: string, postUri: string, pdsEndpoint: string): Promise<PDSDataFetchResult> {
    try {
      // Extract the record key from the URI
      const uriParts = postUri.split('/');
      const collection = uriParts[uriParts.length - 2];
      const rkey = uriParts[uriParts.length - 1];

      // Fetch the post record
      const recordResponse = await fetch(
        `${pdsEndpoint}/xrpc/com.atproto.repo.getRecord?repo=${authorDid}&collection=${collection}&rkey=${rkey}`,
        {
          headers: { 'Accept': 'application/json' },
          signal: AbortSignal.timeout(this.FETCH_TIMEOUT_MS)
        }
      );

      if (!recordResponse.ok) {
        return {
          success: false,
          error: `Record fetch failed: ${recordResponse.status}`
        };
      }

      const recordData = await recordResponse.json();
      
      if (recordData.uri && recordData.cid && recordData.value) {
        // Process the post record
        await eventProcessor.processRecord(
          recordData.uri,
          recordData.cid,
          authorDid,
          recordData.value
        );
        
        console.log(`[PDS_FETCHER] Fetched and processed post ${postUri}`);
        
        return {
          success: true,
          data: { uri: recordData.uri, cid: recordData.cid, record: recordData.value }
        };
      } else {
        return {
          success: false,
          error: 'Record response missing required fields'
        };
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Fetch any record by its AT URI and process it through the event processor
   */
  private async fetchRecordByUri(uri: string, pdsEndpoint: string): Promise<PDSDataFetchResult> {
    try {
      const { repo, collection, rkey } = this.parseAtUri(uri);

      const recordResponse = await fetch(
        `${pdsEndpoint}/xrpc/com.atproto.repo.getRecord?repo=${repo}&collection=${collection}&rkey=${rkey}`,
        {
          headers: { 'Accept': 'application/json' },
          signal: AbortSignal.timeout(this.FETCH_TIMEOUT_MS),
        },
      );

      if (!recordResponse.ok) {
        return { success: false, error: `Record fetch failed: ${recordResponse.status}` };
      }

      const recordData = await recordResponse.json();

      if (recordData.uri && recordData.cid && recordData.value) {
        await eventProcessor.processRecord(recordData.uri, recordData.cid, repo, recordData.value);
        return { success: true, data: recordData };
      }

      return { success: false, error: 'Record response missing required fields' };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Parse AT URI of the form at://did:.../collection/rkey
   */
  private parseAtUri(uri: string): { repo: string; collection: string; rkey: string } {
    const parts = uri.split('/');
    if (parts.length < 5) {
      throw new Error(`Invalid AT URI: ${uri}`);
    }
    // parts: ['at:', '', 'did:plc:...', 'app.bsky.collection', 'rkey']
    const repo = parts[2];
    const collection = parts[3];
    const rkey = parts[4];
    return { repo, collection, rkey };
  }

  /**
   * Get statistics about incomplete entries
   */
  getStats() {
    const stats = {
      total: this.incompleteEntries.size,
      byType: {} as Record<string, number>,
      byRetryCount: {} as Record<string, number>,
      oldestEntry: 0
    };

    let oldestTime = Date.now();

    for (const entry of this.incompleteEntries.values()) {
      // Count by type
      stats.byType[entry.type] = (stats.byType[entry.type] || 0) + 1;
      
      // Count by retry count
      stats.byRetryCount[entry.retryCount] = (stats.byRetryCount[entry.retryCount] || 0) + 1;
      
      // Find oldest entry
      if (entry.lastAttempt < oldestTime) {
        oldestTime = entry.lastAttempt;
      }
    }

    stats.oldestEntry = Date.now() - oldestTime;

    return stats;
  }

  /**
   * Clear all incomplete entries (for testing or manual reset)
   */
  clearAll() {
    const count = this.incompleteEntries.size;
    this.incompleteEntries.clear();
    console.log(`[PDS_FETCHER] Cleared ${count} incomplete entries`);
  }
}

export const pdsDataFetcher = new PDSDataFetcher();

File: python-firehose/pds_data_fetcher.py
#!/usr/bin/env python3
"""
PDS Data Fetcher Service

Fetches missing data from source PDS when entries are incomplete
due to missing referenced users or posts.
Python port of server/services/pds-data-fetcher.ts
"""

import asyncio
import logging
import time
from typing import Optional, Dict, Any, List
from collections import defaultdict
import aiohttp
import asyncpg

from did_resolver import did_resolver

logger = logging.getLogger(__name__)


class IncompleteEntry:
    """Represents an incomplete database entry needing data from PDS"""
    
    def __init__(self, entry_type: str, did: str, uri: Optional[str] = None, missing_data: Any = None):
        self.type = entry_type
        self.did = did
        self.uri = uri
        self.missing_data = missing_data
        self.retry_count = 0
        self.last_attempt = time.time() * 1000


class PDSDataFetcher:
    """Fetches missing data from Personal Data Servers"""
    
    def __init__(self, db_pool):
        self.db_pool = db_pool
        self.max_retry_attempts = 3
        self.retry_delay_ms = 30000  # 30 seconds
        self.fetch_timeout = 10  # seconds
        self.batch_log_size = 5000
        self.incomplete_entries: Dict[str, IncompleteEntry] = {}
        self.is_processing = False
        self.fetch_count = 0
        self.update_count = 0
        self.success_count = 0
        self.post_count = 0
        
        # HTTP session
        self.session: Optional[aiohttp.ClientSession] = None
        
        # Event processor reference (set after initialization)
        self.event_processor = None
        
    async def initialize(self):
        """Initialize HTTP session and start periodic processing"""
        if not self.session:
            self.session = aiohttp.ClientSession()
            
        # Initialize DID resolver
        await did_resolver.initialize()
        
        # Start periodic processing
        asyncio.create_task(self.periodic_processing())
        
    async def close(self):
        """Close HTTP session"""
        if self.session:
            await self.session.close()
            self.session = None
            
        await did_resolver.close()
        
    def sanitize_did(self, did: str) -> str:
        """Sanitize and validate DID format"""
        if not did:
            return ''
            
        original = did
        
        # Remove whitespace
        cleaned = did.replace(' ', '').replace('\t', '').replace('\n', '')
        
        # Remove duplicate colons
        cleaned = ':'.join(part for part in cleaned.split(':') if part)
        
        # Add back 'did:' prefix if it was removed
        if not cleaned.startswith('did:'):
            cleaned = 'did:' + cleaned
            
        # Remove trailing punctuation
        cleaned = cleaned.rstrip(':;,._-')
        
        if cleaned != original:
            logger.warning(f"[PDS_FETCHER] Cleaned malformed DID: '{original}' → '{cleaned}'")
            
        # Validate format
        if not cleaned.startswith('did:'):
            logger.warning(f"[PDS_FETCHER] Invalid DID format: '{cleaned}'")
            
        return cleaned
        
    def mark_incomplete(
        self,
        entry_type: str,
        did: str,
        uri: Optional[str] = None,
        missing_data: Any = None
    ):
        """Mark an entry as incomplete and needing data fetch"""
        clean_did = self.sanitize_did(did)
        key = f"{entry_type}:{clean_did}:{uri}" if uri else f"{entry_type}:{clean_did}"
        
        if key in self.incomplete_entries:
            self.incomplete_entries[key].retry_count += 1
            self.incomplete_entries[key].last_attempt = time.time() * 1000
            self.incomplete_entries[key].missing_data = missing_data or self.incomplete_entries[key].missing_data
        else:
            self.incomplete_entries[key] = IncompleteEntry(entry_type, clean_did, uri, missing_data)
            
    async def periodic_processing(self):
        """Periodically process incomplete entries every 30 seconds"""
        while True:
            await asyncio.sleep(30)
            
            if self.is_processing:
                continue
                
            try:
                await self.process_incomplete_entries()
            except Exception as e:
                logger.error(f"[PDS_FETCHER] Error in periodic processing: {str(e)}")
                
    async def process_incomplete_entries(self):
        """Process all incomplete entries"""
        if not self.incomplete_entries:
            return
            
        self.is_processing = True
        logger.info(f"[PDS_FETCHER] Processing {len(self.incomplete_entries)} incomplete entries...")
        
        entries = list(self.incomplete_entries.items())
        processed = 0
        success = 0
        
        for key, entry in entries:
            try:
                # Skip if too many retries
                if entry.retry_count >= self.max_retry_attempts:
                    logger.warning(f"[PDS_FETCHER] Max retries exceeded for {entry.type} {entry.uri or entry.did}")
                    
                    # For users, create minimal record
                    if entry.type == 'user':
                        try:
                            clean_did = self.sanitize_did(entry.did)
                            handle = await did_resolver.resolve_did_to_handle(clean_did)
                            
                            async with self.db_pool.acquire() as conn:
                                await conn.execute(
                                    """
                                    INSERT INTO users (did, handle, created_at)
                                    VALUES ($1, $2, NOW())
                                    ON CONFLICT (did) DO UPDATE SET handle = EXCLUDED.handle
                                    """,
                                    clean_did, handle or clean_did
                                )
                                
                            logger.info(f"[PDS_FETCHER] Created minimal user record for {clean_did}")
                            
                            # Flush pending operations
                            if self.event_processor:
                                await self.event_processor.flush_pending_user_ops(conn, clean_did)
                        except Exception as e:
                            logger.error(f"[PDS_FETCHER] Failed to create minimal user for {entry.did}: {str(e)}")
                            
                    del self.incomplete_entries[key]
                    continue
                    
                # Skip if recently attempted
                time_since_last = (time.time() * 1000) - entry.last_attempt
                if time_since_last < self.retry_delay_ms:
                    continue
                    
                # Fetch missing data
                result = await self.fetch_missing_data(entry)
                if result['success']:
                    success += 1
                    del self.incomplete_entries[key]
                    self.success_count += 1
                    
                    if self.success_count % self.batch_log_size == 0:
                        logger.info(f"[PDS_FETCHER] {self.batch_log_size} successful fetches (total: {self.success_count})")
                else:
                    logger.warning(f"[PDS_FETCHER] Failed to fetch {entry.type} {entry.uri or entry.did}: {result.get('error')}")
                    
                processed += 1
            except Exception as e:
                logger.error(f"[PDS_FETCHER] Error processing {entry.type} {entry.uri or entry.did}: {str(e)}")
                
        logger.info(f"[PDS_FETCHER] Processed {processed} entries, {success} successful, {len(self.incomplete_entries)} remaining")
        self.is_processing = False
        
    async def fetch_missing_data(self, entry: IncompleteEntry) -> Dict[str, Any]:
        """Fetch missing data from PDS"""
        try:
            clean_did = self.sanitize_did(entry.did)
            
            # Validate DID format
            if not (clean_did.startswith('did:plc:') or clean_did.startswith('did:web:')):
                return {'success': False, 'error': f'Invalid DID format: {clean_did}'}
                
            # Resolve DID to PDS endpoint
            pds_endpoint = await did_resolver.resolve_did_to_pds(clean_did)
            if not pds_endpoint:
                return {'success': False, 'error': f'Could not resolve PDS endpoint for DID: {clean_did}'}
                
            # Batch logging
            self.fetch_count += 1
            if self.fetch_count % self.batch_log_size == 0:
                logger.info(f"[PDS_FETCHER] Fetched data for {self.batch_log_size} entries (total: {self.fetch_count})")
                
            # Route to appropriate handler
            if entry.type == 'user':
                return await self.fetch_user_data(clean_did, pds_endpoint)
            elif entry.type == 'post':
                return await self.fetch_post_data(clean_did, entry.uri, pds_endpoint)
            elif entry.type in ['list', 'listitem', 'feedgen', 'starterpack', 'labeler', 'record']:
                return await self.fetch_record_by_uri(entry.uri, pds_endpoint)
            elif entry.type in ['like', 'repost', 'follow']:
                # Ensure the actor exists first
                return await self.fetch_user_data(clean_did, pds_endpoint)
            else:
                return {'success': False, 'error': f'Unknown entry type: {entry.type}'}
        except Exception as e:
            return {'success': False, 'error': str(e)}
            
    async def fetch_user_data(self, did: str, pds_endpoint: str) -> Dict[str, Any]:
        """Fetch user profile data from PDS"""
        try:
            encoded_did = aiohttp.helpers.quote(did, safe='')
            url = f"{pds_endpoint}/xrpc/com.atproto.repo.getRecord?repo={encoded_did}&collection=app.bsky.actor.profile&rkey=self"
            
            timeout = aiohttp.ClientTimeout(total=self.fetch_timeout)
            async with self.session.get(url, headers={'Accept': 'application/json'}, timeout=timeout) as response:
                if response.status == 400:
                    text = await response.text()
                    if 'RecordNotFound' in text:
                        # Account exists but has no profile
                        handle = await did_resolver.resolve_did_to_handle(did)
                        
                        async with self.db_pool.acquire() as conn:
                            await conn.execute(
                                """
                                INSERT INTO users (did, handle, created_at)
                                VALUES ($1, $2, NOW())
                                ON CONFLICT (did) DO UPDATE SET handle = EXCLUDED.handle
                                """,
                                did, handle or did
                            )
                            
                        logger.warning(f"[PDS_FETCHER] No profile record at PDS for {did} - created minimal user")
                        
                        # Flush pending operations
                        if self.event_processor:
                            async with self.db_pool.acquire() as conn:
                                await self.event_processor.flush_pending_user_ops(conn, did)
                                
                        return {'success': True, 'data': {'did': did, 'handle': handle or did, 'profile': None}}
                        
                if response.status != 200:
                    error_text = await response.text()
                    return {'success': False, 'error': f'Profile fetch failed: {response.status} - {error_text[:200]}'}
                    
                data = await response.json()
                profile = data.get('value')
                
                if profile:
                    handle = await did_resolver.resolve_did_to_handle(did)
                    
                    # Extract avatar and banner CIDs
                    avatar_cid = self.extract_blob_cid(profile.get('avatar'))
                    banner_cid = self.extract_blob_cid(profile.get('banner'))
                    
                    # Update user with full profile
                    async with self.db_pool.acquire() as conn:
                        await conn.execute(
                            """
                            INSERT INTO users (did, handle, display_name, description, avatar_url, banner_url, created_at)
                            VALUES ($1, $2, $3, $4, $5, $6, NOW())
                            ON CONFLICT (did) DO UPDATE SET
                                handle = EXCLUDED.handle,
                                display_name = EXCLUDED.display_name,
                                description = EXCLUDED.description,
                                avatar_url = EXCLUDED.avatar_url,
                                banner_url = EXCLUDED.banner_url
                            """,
                            did, handle or did, profile.get('displayName'), profile.get('description'), avatar_cid, banner_cid
                        )
                        
                    self.update_count += 1
                    if self.update_count % self.batch_log_size == 0:
                        logger.info(f"[PDS_FETCHER] Updated {self.batch_log_size} users (total: {self.update_count})")
                        
                    # Flush pending operations
                    if self.event_processor:
                        async with self.db_pool.acquire() as conn:
                            await self.event_processor.flush_pending_user_ops(conn, did)
                            
                    return {'success': True, 'data': {'did': did, 'handle': handle or did, 'profile': profile}}
                else:
                    return {'success': False, 'error': 'No profile record found'}
        except asyncio.TimeoutError:
            return {'success': False, 'error': 'Timeout'}
        except Exception as e:
            return {'success': False, 'error': str(e)}
            
    async def fetch_post_data(self, author_did: str, post_uri: str, pds_endpoint: str) -> Dict[str, Any]:
        """Fetch post data from PDS"""
        try:
            # Extract collection and rkey from URI
            uri_parts = post_uri.split('/')
            collection = uri_parts[-2]
            rkey = uri_parts[-1]
            
            # URL encode parameters
            encoded_did = aiohttp.helpers.quote(author_did, safe='')
            encoded_collection = aiohttp.helpers.quote(collection, safe='')
            encoded_rkey = aiohttp.helpers.quote(rkey, safe='')
            
            url = f"{pds_endpoint}/xrpc/com.atproto.repo.getRecord?repo={encoded_did}&collection={encoded_collection}&rkey={encoded_rkey}"
            
            timeout = aiohttp.ClientTimeout(total=self.fetch_timeout)
            async with self.session.get(url, headers={'Accept': 'application/json'}, timeout=timeout) as response:
                if response.status in [400, 404]:
                    text = await response.text()
                    if 'RecordNotFound' in text:
                        logger.warning(f"[PDS_FETCHER] Post not found (deleted): {post_uri}")
                        return {'success': True, 'data': None}  # Treat as success to stop retrying
                        
                if response.status != 200:
                    error_text = await response.text()
                    return {'success': False, 'error': f'Record fetch failed: {response.status} - {error_text[:200]}'}
                    
                record_data = await response.json()
                
                if record_data.get('uri') and record_data.get('cid') and record_data.get('value'):
                    # Process the post record via event processor
                    if self.event_processor:
                        await self.event_processor.process_record(
                            record_data['uri'],
                            record_data['cid'],
                            author_did,
                            record_data['value']
                        )
                        
                    self.post_count += 1
                    if self.post_count % self.batch_log_size == 0:
                        logger.info(f"[PDS_FETCHER] Fetched and processed {self.batch_log_size} posts (total: {self.post_count})")
                        
                    return {'success': True, 'data': record_data}
                else:
                    return {'success': False, 'error': 'Record response missing required fields'}
        except asyncio.TimeoutError:
            return {'success': False, 'error': 'Timeout'}
        except Exception as e:
            return {'success': False, 'error': str(e)}
            
    async def fetch_record_by_uri(self, uri: str, pds_endpoint: str) -> Dict[str, Any]:
        """Fetch any record by AT URI"""
        try:
            repo, collection, rkey = self.parse_at_uri(uri)
            
            # URL encode parameters
            encoded_repo = aiohttp.helpers.quote(repo, safe='')
            encoded_collection = aiohttp.helpers.quote(collection, safe='')
            encoded_rkey = aiohttp.helpers.quote(rkey, safe='')
            
            url = f"{pds_endpoint}/xrpc/com.atproto.repo.getRecord?repo={encoded_repo}&collection={encoded_collection}&rkey={encoded_rkey}"
            
            timeout = aiohttp.ClientTimeout(total=self.fetch_timeout)
            async with self.session.get(url, headers={'Accept': 'application/json'}, timeout=timeout) as response:
                if response.status in [400, 404]:
                    text = await response.text()
                    if 'RecordNotFound' in text:
                        logger.warning(f"[PDS_FETCHER] Record not found (deleted): {uri}")
                        return {'success': True, 'data': None}
                        
                if response.status != 200:
                    error_text = await response.text()
                    return {'success': False, 'error': f'Record fetch failed: {response.status} - {error_text[:200]}'}
                    
                record_data = await response.json()
                
                if record_data.get('uri') and record_data.get('cid') and record_data.get('value'):
                    # Process via event processor
                    if self.event_processor:
                        await self.event_processor.process_record(
                            record_data['uri'],
                            record_data['cid'],
                            repo,
                            record_data['value']
                        )
                        
                    return {'success': True, 'data': record_data}
                else:
                    return {'success': False, 'error': 'Record response missing required fields'}
        except asyncio.TimeoutError:
            return {'success': False, 'error': 'Timeout'}
        except Exception as e:
            return {'success': False, 'error': str(e)}
            
    def parse_at_uri(self, uri: str) -> tuple:
        """Parse AT URI of form at://did:.../collection/rkey"""
        parts = uri.split('/')
        if len(parts) < 5:
            raise ValueError(f"Invalid AT URI: {uri}")
            
        # parts: ['at:', '', 'did:plc:...', 'app.bsky.collection', 'rkey']
        repo = parts[2]
        collection = parts[3]
        rkey = parts[4]
        return repo, collection, rkey
        
    def extract_blob_cid(self, blob: Any) -> Optional[str]:
        """Extract CID from blob reference"""
        if not blob:
            return None
            
        if isinstance(blob, str):
            return None if blob == 'undefined' else blob
            
        if isinstance(blob, dict):
            ref = blob.get('ref')
            if isinstance(ref, str):
                return None if ref == 'undefined' else ref
            elif isinstance(ref, dict):
                link = ref.get('$link')
                if link:
                    return None if link == 'undefined' else link
                # Handle binary CID (would need additional processing)
                if 'code' in ref and 'multihash' in ref:
                    # For simplicity, try to call toString if it's an object
                    if hasattr(ref, '__str__'):
                        return str(ref)
                        
            cid = blob.get('cid')
            if cid:
                return None if cid == 'undefined' else cid
                
        return None
        
    def get_stats(self) -> Dict[str, Any]:
        """Get statistics about incomplete entries"""
        stats = {
            'total': len(self.incomplete_entries),
            'by_type': {},
            'by_retry_count': {},
            'oldest_entry': 0
        }
        
        oldest_time = time.time() * 1000
        
        for entry in self.incomplete_entries.values():
            # Count by type
            stats['by_type'][entry.type] = stats['by_type'].get(entry.type, 0) + 1
            
            # Count by retry count
            stats['by_retry_count'][entry.retry_count] = stats['by_retry_count'].get(entry.retry_count, 0) + 1
            
            # Find oldest
            if entry.last_attempt < oldest_time:
                oldest_time = entry.last_attempt
                
        stats['oldest_entry'] = (time.time() * 1000) - oldest_time
        
        return stats
